### PR TITLE
feat: added custom status code for a missed block

### LINKED ISSUE #12399: RPC feature: new error code for missing blocks
### Description

We have a request from one of our partners to make specific 4** error (now it's 404 not found) for blocks that do not exist but are between existing blocks

### PR DESCRIPTION
**Pull request by frolvanya**
As requested from our partner, custom status code for a missing block was added. I decided to go with 422 Unprocessable Content which seems pretty logical to me, but let me know if this is a bad choice

closes #12399

### CODE REVIEW

**Comment by walnut-the-cat** [comment]
hmm is UNKNOWN_BLOCK right one? From what i know, in nearcore, UNKNOWN_CHUNK is used when a node receives a chunk for the future block. So in that sense, UNKNOWN_BLOCK sounds like something similar (e.g. orphaned block). Maybe the meaning is different with RPC?

**Comment by frolvanya** [comment]
> hmm is UNKNOWN_BLOCK right one? From what i know, in nearcore, UNKNOWN_CHUNK is used when a node receives a chunk for the future block. So in that sense, UNKNOWN_BLOCK sounds like something similar (e.g. orphaned block). Maybe the meaning is different with RPC?

I mean "UNKNOWN_BLOCK" is a part of a handler error which is being returned when input block is either too low or too high 
image

I tried to manually find a "missing" block to test it on, but I was out of luck

**Comment by nhovsmith-cb** [comment]
@frolvanya  From  [explorer:block#131,402,418](https://nearblocks.io/blocks/gpAxKrtzHUq4TcWUPVQEuexjRnaBBWdNZVyuRuQ4boM) observe that the parent is [explorer:block#131,402,416](https://nearblocks.io/blocks/2Xnng59vbAXZxEL8wNTmN6v7znmsxugm2rHjjRpnCpZE). The network seemed to have skipped # 131,402,417

**Comment by frolvanya** [comment]
> @frolvanya From [explorer:block#131,402,418](https://nearblocks.io/blocks/gpAxKrtzHUq4TcWUPVQEuexjRnaBBWdNZVyuRuQ4boM) observe that the parent is [explorer:block#131,402,416](https://nearblocks.io/blocks/2Xnng59vbAXZxEL8wNTmN6v7znmsxugm2rHjjRpnCpZE). The network seemed to have skipped # 131,402,417

Thanks for the info! Then I have 2 updates:
1) Error is correct ("UNKNOWN_BLOCK")
2) `"cause"."info"` field is empty for some reason and that means that my logic won't work there. I'll try to modify current implementation tomorrow

**Comment by bowenwang1996** [comment]
@frolvanya could you add a unit test so that we know if something breaks it in the future?

**Comment by frolvanya** [comment]
> @frolvanya could you add a unit test so that we know if something breaks it in the future?

Sure, I'm currently working on it

**Comment by xingyaow** [comment]
Hi team, can we get an ETA of when will this PR gets merged? Currently waiting for this new status code to unblock us, many thanks!

**Comment by walnut-the-cat** [comment]
Hi @xingyaow, sine this is nearcore fix, even when it is merged, it won't be automatically deployed. It needs to go through protocol version upgrade (either major or minor). By default, after the PR is merged in, it becomes candidate for 2.4 release, which is scheduled in mid December. 

If it is very urgent from your end, we can try to prepare upgrade and release early next week. 

Let us know-

**Comment by xingyaow** [comment]
Thanks, @walnut-the-cat! Yes it's urgent from our side as it's blocking us to launch Near by the end of December. We can't wait til mid December as we will have code freeze in December. Much appreciate it if this can be deployed and released sooner!

**Comment by walnut-the-cat** [comment]
@xingyaow ,

We will try that. In the meantime, this PR is basically code complete and missing part is thorough testing. @frolvanya thinks this should not cause any major issue even in the worst case, since it does not change any RPC logic. 

If you are in hurry, it should be fairly safe for you to apply the PR on your RPC and unblock yourself for timebeing  (Then later replace with proper binary)

File: chain/jsonrpc/CHANGELOG.md
Comment (walnut-the-cat) [inline chain/jsonrpc/CHANGELOG.md:3]: 2.3.1

File: pytest/tests/sanity/rpc_missing_block.py
Comment (Trisfald) [inline pytest/tests/sanity/rpc_missing_block.py:74]: Perhaps you meant
```suggestion
        assert False, "Expected an exception for block height {height} but none was raised"
```

File: pytest/tests/sanity/rpc_missing_block.py
Comment (Trisfald) [inline pytest/tests/sanity/rpc_missing_block.py:68]: Why is killing node 1 necessary?

**Review by Trisfald — APPROVED** [review]

File: pytest/tests/sanity/rpc_missing_block.py
Comment (frolvanya) [reply pytest/tests/sanity/rpc_missing_block.py:68]: In order to replicate a missing block state we set up nodes and then kill one. After this procedure many blocks will be considered as missing. This idea was proposed by @marcelo-gonzalez and it works great locally

File: chain/jsonrpc/CHANGELOG.md
Comment (frolvanya) [reply chain/jsonrpc/CHANGELOG.md]: Fixed in 7070960

File: chain/jsonrpc/src/lib.rs
Comment (frolvanya) [reply chain/jsonrpc/src/lib.rs]: Fixed in 602789d5f8ef3d1fbb47e57e04ffe9909196047e

File: chain/jsonrpc/src/lib.rs
Comment (frolvanya) [reply chain/jsonrpc/src/lib.rs:1398]: Sure, it would be great to improve RPC error structures in the future

**Pull request merged**
